NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -1,45 +1,143 @@
+quantum_dimension(::AbelianGroup, ::AbstractCategory) = 1
+quantum_dimension(::EmptyCategory, ::AbstractCategory) = 0
+
+function quantum_dimension(::SymmetryStyle, g::AbstractUnitRange)
+  gblocks = BlockArrays.blocklengths(g)
+  return sum(gblocks .* quantum_dimension.(LabelledNumbers.label.(gblocks)))
 end
 
+quantum_dimension(::AbelianGroup, g::AbstractUnitRange) = length(g)
+function quantum_dimension(::SymmetryStyle, g::GradedAxes.UnitRangeDual)
+  return quantum_dimension(GradedAxes.dual(g))
+end
+quantum_dimension(::AbelianGroup, g::GradedAxes.UnitRangeDual) = length(g)  # resolves ambiguity
+
+# ================  fusion rule interface ====================
 ⊗(c1::AbstractCategory, c2::AbstractCategory) = fusion_rule(c1, c2)
 
-⊕(c1::AbstractCategory, c2::AbstractCategory) = [c1, c2]
-⊕(cs::Vector{<:AbstractCategory}, c::AbstractCategory) = [cs; c]
-⊕(c::AbstractCategory, cs::Vector{<:AbstractCategory}) = [c; cs]
+function fusion_rule(c1, c2)
+  return fusion_rule(combine_styles(SymmetryStyle(c1), SymmetryStyle(c2)), c1, c2)
+end
 
-function Base.show(io::IO, cs::Vector{<:AbstractCategory})
-  (length(cs) <= 1) && print(io, "[")
-  symbol = ""
-  for c in cs
-    print(io, symbol, c)
-    symbol = " ⊕ "
-  end
-  (length(cs) <= 1) && print(io, "]")
-  return nothing
+function fusion_rule(::SymmetryStyle, c1::C, c2::C) where {C<:AbstractCategory}
+  degen, labels = label_fusion_rule(C, category_label(c1), category_label(c2))
+  return GradedAxes.gradedrange(LabelledNumbers.LabelledInteger.(degen, C.(labels)))
 end
 
-function trivial(category_type::Type{<:AbstractCategory})
-  return error("`trivial` not defined for type $(category_type).")
+# abelian case: return Category
+function fusion_rule(::AbelianGroup, c1::C, c2::C) where {C<:AbstractCategory}
+  return C(label_fusion_rule(C, category_label(c1), category_label(c2)))
 end
 
-istrivial(c::AbstractCategory) = (c == trivial(typeof(c)))
+function label_fusion_rule(category_type::Type{<:AbstractCategory}, l1, l2)
+  return error("`label_fusion_rule` not defined for type $(category_type).")
+end
 
-function dual(category_type::Type{<:AbstractCategory})
-  return error("`dual` not defined for type $(category_type).")
+# convenient to define fusion rule for LabelledInteger too
+# TBD expose this through ⊗? Currently not accessible.
+function fusion_rule(
+  ::SymmetryStyle, l1::LabelledNumbers.LabelledInteger, l2::LabelledNumbers.LabelledInteger
+)
+  blocks12 = BlockArrays.blocklengths(LabelledNumbers.label(l1) ⊗ LabelledNumbers.label(l2))
+  v =
+    LabelledNumbers.LabelledInteger.(l1 * l2 .* blocks12, LabelledNumbers.label.(blocks12))
+  return GradedAxes.gradedrange(v)
+end
+
+function fusion_rule(
+  ::AbelianGroup, l1::LabelledNumbers.LabelledInteger, l2::LabelledNumbers.LabelledInteger
+)
+  fused = LabelledNumbers.label(l1) ⊗ LabelledNumbers.label(l2)
+  return LabelledNumbers.LabelledInteger(l1 * l2, fused)
```

Comment:
This goes back to our discussion about what `fusion_rule` should be used for. My opinion is that it should be considered very low-level, so maybe only defined for categories, and anything up above that level that can be derived from the low-level `fusion_rule` definitions should be called something else, maybe `fusion_product`.

The analogy I have in mind is how Julia's type promotion system works: https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#Promotion where the external interface is based around `promote`/`promote_type` and there are low-level promotion rules defined through `promote_rule`.

Reply:
Indeed I did not change the fusion rule interface after our last discussion on it. I wanted to see how I would be using it in FusionTensors first before coming back to it, but it is on my todo list :)

Reply:
Makes sense, thanks for the update.

Reply:
Fixed, now using `fuse_blocklengths`.